### PR TITLE
feat: review links actions permissions fixed

### PR DESCRIPTION
**Pull request by intsco**
Key changes in this PR:
* Remove `publication_status` field from the `DatasetProject` model.
* Throw structured errors `{type: 'error_type', message: 'Description'}` on actions related to project publication and subsequent modifications to the project or datasets in it.
* Substantially extend the tests suite for publication related actions.

### CODE REVIEW

File: metaspace/graphql/src/modules/project/controller/Mutation.ts [hunk 1]
```diff
@@ -248,28 +246,29 @@ const MutationResolvers: FieldResolversFor<Mutation, void> = {
...
   async publishProject(source, { projectId }, ctx: Context): Promise<ProjectSource> {
     await asyncAssertCanEditProject(ctx, projectId);
 
-    await ctx.entityManager.update(ProjectModel, projectId,
-      { publicationStatus: PSO.PUBLISHED, isPublic: true });
-    const datasetProjectRepository = await ctx.entityManager.getRepository(DatasetProjectModel);
-    datasetProjectRepository.update({ projectId }, { publicationStatus: PSO.PUBLISHED });
+    await ctx.entityManager.update(ProjectModel, projectId, {
+      publicationStatus: PSO.PUBLISHED,
+      // TODO: publishedDT: utc(),
```
Comment (richardgoater) [inline metaspace/graphql/src/modules/project/controller/Mutation.ts:263]: Think I'm relying on this for the project list page, what happened to it?

Comment (intsco) [reply metaspace/graphql/src/modules/project/controller/Mutation.ts:263]: I didn't want to mess up with additional migrations locally when rebasing onto master, so I just commented it out. You're free to undo this change of course.

Comment (richardgoater) [reply metaspace/graphql/src/modules/project/controller/Mutation.ts:263]: cool, no worries 👍

**Comment by LachlanStuart** [comment]
Btw, no need to use them here, but just so you're aware of them, the `createBackgroundData` and `validateBackgroundData` test utils were made for catching cases of missing `where` clauses causing changes to unrelated data.

File: metaspace/graphql/src/tests/testDataCreation.ts [hunk 2]
```diff
@@ -105,14 +105,13 @@ export const createTestDataset = async (dataset: Partial<Dataset> = {}, engineDa
...
+  const datasetProjectPromise = testEntityManager.save(DatasetProject, {
     projectId: project.id,
     datasetId: dataset.id,
-    approved: true,
-    publicationStatus
-  });
-  return dataset;
+    approved: true
+  } as Partial<DatasetProject>);
+  return (await datasetProjectPromise) as DatasetProject;
```
Comment (intsco) [inline metaspace/graphql/src/tests/testDataCreation.ts:116]: @LachlanStuart this is the best I managed to come up with.

**Pull request merged**